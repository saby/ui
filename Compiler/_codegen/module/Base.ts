/**
 */

import {
    IDependenciesController,
    createDependenciesController,
} from './Dependencies';
import { IExportsController, createExportsController } from './Exports';

/**
 * Интерфейс процессора, который выполняет компиляцию модуля в различный формат.
 */
export interface IModuleProcessor {
    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void;

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void;

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void;

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void;

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void;

    /**
     * Выполнить компиляцию модуля.
     * @returns {string} JavaScript код модуля.
     */
    compile(): string;
}

/**
 * Абстрактный базовый класс процессора модулей.
 * Предоставляет общие данные и методы для выполнения компиляции.
 */
export abstract class Base implements IModuleProcessor {
    /**
     * Имя компилируемого модуля.
     * @protected
     */
    protected name: string;

    /**
     * Флаг включения строго режима для компилируемого модуля.
     * @protected
     */
    protected useStrict: boolean;

    /**
     * Массив кода, который составляет тело будущего модуля.
     * @protected
     */
    protected code: string[];

    /**
     * Инстанс контроллера зависимостей, необходимый для работы с именованными и
     * безымянными (анонимными) зависимостями модуля.
     * @protected
     */
    protected dependenciesController: IDependenciesController;

    /**
     * Инстанс контроллера экспортируемых данных.
     * @protected
     */
    protected exportsController: IExportsController;

    /**
     * Инициализировать инстанс базового класса.
     * @protected
     */
    protected constructor() {
        this.name = '';
        this.useStrict = false;
        this.code = [];
        this.dependenciesController = createDependenciesController();
        this.exportsController = createExportsController();
    }

    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void {
        this.name = name;
    }

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void {
        this.useStrict = value;
    }

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void {
        this.dependenciesController.addDependency(ref, name);
    }

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void {
        this.code.push(block);
    }

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void {
        this.exportsController.setReturnableExport(block);
    }

    /**
     * Выполнить компиляцию модуля.
     * @abstract
     * @returns {string} JavaScript код модуля.
     */
    abstract compile(): string;
}
