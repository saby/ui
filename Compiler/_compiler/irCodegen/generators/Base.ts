/**
 * @author Krylov M.A.
 *
 * Базовый генератор обращений к именованным сущностям.
 */

import type { TPrimitive, IFormatter } from '../Interface';
import type { TAliasesMap } from '../Aliases';

import { wrapSequence } from '../types/String';

export default class Base {
    protected constructor(
        protected readonly formatter: IFormatter,
        protected readonly instance: string,
        protected readonly aliases: TAliasesMap,
        protected readonly isRelease: boolean
    ) { }

    protected getPropertyName(fn: string): string {
        if (this.isRelease) {
            if (this.aliases.has(fn)) {
                return `.${this.aliases.get(fn)}/*${fn}*/`;
            }
        }

        return `.${fn}`;
    }

    protected generateMethodCall(method: string, args: TPrimitive[]): string {
        return `${this.instance}${this.getPropertyName(method)}(${wrapSequence(args)})`;
    }

    protected generateFormattedMethodCall(method: string, args: TPrimitive[]): string {
        return `${this.instance}${this.getPropertyName(method)}(${this.formatter.formatSequence(args, 1)})`;
    }
}
