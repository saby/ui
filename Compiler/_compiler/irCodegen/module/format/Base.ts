/**
 * @author Krylov M.A.
 */

import type { IModuleGenerator, IDependenciesController, IExportsController } from '../Interface';

/**
 * Абстрактный базовый класс процессора модулей.
 * Предоставляет общие данные и методы для выполнения компиляции.
 *
 * @private
 */
export default abstract class Base implements IModuleGenerator {

    /**
     * Имя компилируемого модуля.
     * @protected
     */
    protected name: string;

    /**
     * Флаг включения строго режима для компилируемого модуля.
     * @protected
     */
    protected useStrict: boolean;

    /**
     * Содержимое модуля.
     * @protected
     */
    protected contents: string[];

    /**
     * Инстанс контроллера зависимостей, необходимый для работы с именованными и
     * безымянными (анонимными) зависимостями модуля.
     * @protected
     */
    protected dependenciesController: IDependenciesController;

    /**
     * Инстанс контроллера экспортируемых данных.
     * @protected
     */
    protected exportsController: IExportsController;

    /**
     * Инициализировать инстанс базового класса.
     * @protected
     * @param {IDependenciesController} dependencies Контроллер зависимостей.
     * @param {IExportsController} exports Контроллер экспортируемых данных.
     * @protected
     */
    constructor(dependencies: IDependenciesController, exports: IExportsController) {
        this.name = '';
        this.useStrict = false;
        this.contents = [];

        this.dependenciesController = dependencies;
        this.exportsController = exports;
    }

    /**
     * Установить имя компилируемого модуля.
     * @param {string} name Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void {
        this.name = name;
    }

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param {boolean} value Значение флага строго режима.
     */
    setStrictMode(value: boolean): void {
        this.useStrict = value;
    }

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param {string} ref Путь зависимости.
     * @param {string?} name Имя зависимости.
     */
    addDependency(ref: string, name?: string): void {
        this.dependenciesController.addDependency(ref, name);
    }

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param {string} block Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void {
        this.contents.push(block);
    }

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param {string} block Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void {
        this.exportsController.setReturnableExport(block);
    }

    /**
     * Выполнить компиляцию модуля.
     * @abstract
     * @returns {string} JavaScript код модуля.
     */
    abstract compile(): string;
}
