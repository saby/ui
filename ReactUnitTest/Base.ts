import { Control } from 'UICore/Base';
import { IControlOptions } from 'UICommon/Base';

export interface ITestOptions extends IControlOptions {
    testName?: string;
    afterMountCallback?: Function;
    afterUpdateCallback?: Function;
}

export class TestBaseControl<
    IOptions extends ITestOptions = ITestOptions
> extends Control<IOptions> {
    testName: string = '';
    afterMountCallback: Function = null;
    afterUpdateCallback: Function = null;
    _beforeMount(options: ITestOptions): void {
        this.afterMountCallback = options.afterMountCallback;
        this.afterUpdateCallback = options.afterUpdateCallback;
        this.testName = options.testName;
    }
    _afterMount(): void {
        this.afterMountCallback?.(this);
    }
    _afterUpdate(): void {
        this.afterUpdateCallback?.(this);
    }
}
