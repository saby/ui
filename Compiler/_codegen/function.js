define('Compiler/_codegen/function', [
   'Compiler/_utils/Helpers',
   'Compiler/_expressions/Process',
   'Compiler/_expressions/Event',
   'Compiler/_modules/utils/common',
   'Compiler/_modules/if',
   'Compiler/_modules/for',
   'Compiler/_modules/else',
   'Compiler/_modules/partial',
   'Compiler/_modules/template',
   'Compiler/_modules/utils/tag',
   'Compiler/_modules/data/utils/functionStringCreator',
   'Compiler/_utils/ErrorHandler',
   'Compiler/_codegen/JsTemplates',
   'Compiler/_codegen/Generator',
   'Compiler/_codegen/TClosure',
   'Compiler/Config',
   'Compiler/_codegen/Template',
   'Compiler/_codegen/ECMAScript'
], function processingModule(
   Helpers,
   Process,
   eventExpressions,
   utils,
   ifModule,
   forModule,
   elseModule,
   partialModule,
   templateModule,
   tagUtils,
   FSC,
   ErrorHandlerLib,
   JsTemplates,
   Generator,
   TClosure,
   builderConfig,
   TemplateLib,
   ECMAScriptLib
) {
   'use strict';

   var EMPTY_STRING = '';

   var errorHandler = ErrorHandlerLib.createErrorHandler(true);

   function createAttrObject(val) {
      return {
         type: 'text',
         value: val
      };
   }

   function getFuncNameByFile(fileName) {
      return (
         fileName &&
         fileName.replace &&
         fileName

            // Remove require substitution
            .replace(/^(wml|tmpl|html)!/g, EMPTY_STRING)

            // Remove file extension
            .replace(/\.(wml|tmpl|html)$/gi, EMPTY_STRING)

            // Replace path separator and spaces with underscore
            .replace(/[.\-:/\s]/gi, '_')

            // Remove leading digits at the start
            .replace(/^\d+/gi, EMPTY_STRING)

            // Remove any invalid characters
            .replace(/[^\d\w_]/gi, EMPTY_STRING)
      );
   }

   function isFunctionNameConfigurable(func) {
      if (typeof func !== 'function') {
         return false;
      }

      // Не определять новое имя для функции, если дескриптор этого не позволяет
      // В FF36 <#name>configurable/writable есть false
      var descriptor = Object.getOwnPropertyDescriptor(func, 'name');
      if (!descriptor) {
         return false;
      }
      return descriptor.configurable;
   }

   var processing = {
      name: 'string',

      /**
       * Модули шаблонизатора
       */
      _modules: {
         if: ifModule,
         for: forModule,
         else: elseModule,
         partial: partialModule,
         template: templateModule
      },

      /**
       * Модули, использумые в управляющих атрибутах
       */
      _attributeModules: {
         if: ifModule,
         for: forModule
      },

      /**
       * Данные контролов
       */
      _controlsData: {},
      handlers: {},
      getFuncNameByFile: getFuncNameByFile,
      childrenStorage: [],
      getFuncNameByTemplate: function (wsTemplateName) {
         if (typeof wsTemplateName === 'string') {
            return this.getFuncNameByFile(wsTemplateName);
         }
         if (
            typeof wsTemplateName === 'object' &&
            wsTemplateName.data.type === 'text'
         ) {
            return this.getFuncNameByFile(wsTemplateName.data.value);
         }
         return undefined;
      },
      getFuncName: function getFuncName(
         propertyName,
         fileName,
         wsTemplateName
      ) {
         var fnByTmpl = this.getFuncNameByTemplate(wsTemplateName);
         var fnByFile = this.getFuncNameByFile(fileName);
         var functionName =
            fnByTmpl ||
            fnByFile ||
            propertyName ||
            builderConfig.Config.privateFunctionName;

         // Запомнить вычисленное имя функции и определить ее идентификатор в случае повторения имени
         if (this.functionNames) {
            var isAlreadyExist =
               this.functionNames.hasOwnProperty(functionName);
            if (!isAlreadyExist) {
               this.functionNames[functionName] = 1;
               return functionName;
            }
            var idx = this.functionNames[functionName].toString();
            this.functionNames[functionName]++;
            functionName += '_' + idx;
         }

         // Перед возвратом имени убедиться, что не происходит переопределение
         if (this.functionNames.hasOwnProperty(functionName)) {
            return this.getFuncName(functionName);
         }
         return functionName;
      },
      setFunctionName: function (func, wsTemplateName, fileName, propertyName) {
         // Не определять новое имя для функции, если оно уже существует
         // В IE Отсутствует Function.prototype.name, поэтому здесь необходимо проверять, чтобы name был опеределен
         if (
            typeof func === 'function' &&
            func.name !== 'anonymous' &&
            func.name !== undefined
         ) {
            return func.name;
         }

         if (!isFunctionNameConfigurable(func)) {
            return this.getFuncName(builderConfig.Config.privateFunctionName);
         }

         var functionName = this.getFuncName(
            propertyName,
            fileName,
            wsTemplateName
         );

         // В случае, если пришла функция, то нужно определить ее имя
         if (typeof func === 'function') {
            Object.defineProperty(func, 'name', {
               value: functionName,
               configurable: true
            });
            return func.name;
         }
         return functionName;
      },

      /**
       * Если не чистить данные контролов, на препроцессоре они будут копиться и
       * для всех пользователей будут приходить лишние данные и контролы,
       * которых нет на странице, будут пытаться зареквайриться
       */
      clearControlsData: function () {
         this._controlsData = {};
      },

      /**
       * Декорирование рутового узла
       */
      decorate: function decorate(attributes) {
         return function decorateRoot(rootAttribs) {
            var currentRootAttribs = rootAttribs || {};
            var attrs = Helpers.shallowClone(currentRootAttribs);
            if (attributes) {
               for (var name in attributes) {
                  if (attrs[name]) {
                     if (attrs[name].data.length > 0) {
                        attrs[name].data.push(
                           createAttrObject(' ' + attributes[name])
                        );
                     } else if (attrs[name].data && !attrs[name].data.length) {
                        attrs[name].data = [
                           attrs[name].data,
                           createAttrObject(' ' + attributes[name])
                        ];
                     } else {
                        attrs[name] = {};
                        attrs[name].data = createAttrObject(attributes[name]);
                     }
                  } else {
                     attrs[name] = {};
                     attrs[name].data = createAttrObject(attributes[name]);
                  }
               }
            }
            return attrs;
         };
      },

      /**
       * Получение результирущего объекта
       * @param ast {Array} Список AST узлов.
       * @param data {Object} Данные компиляции.
       * @param handlers {object} Конфигурация компиляции.
       * @param attributes {object} Атрибуты узла.
       * @param appendHeader Флаг, означающий, что необходимо включить заголовок с переменными.
       */
      getString: function getString(
         ast,
         data,
         handlers,
         attributes,
         appendHeader
      ) {
         var decor = this.decorate(attributes);

         /**
          * Нам нужно пометить эту функцию, что она генерирует атрибуты для КОРНЕВОГО тега
          * тогда если корневой тег - partial, то есть, эта функция участвует при построении
          * функции внутреннего шаблона, то мы должны замержить еще и те атрибуты, которые
          * были переданы в сгенерированную функцию с родительского шаблона.
          * То есть контрол
          * MyButton имеет внутри себя корнем ws:partial
          * он же в свою очередь создает div
          * Тогда нужно чтобы при создании <MyButton attr:class="привет" />
          * атрибут class долетел до дива
          */
         decor.isMainAttrs = true;
         if (handlers) {
            /**
             * Конфиги переданные в requirejs плагине для шаблонизатора
             * @type {Array|*}
             */
            this.handlers = handlers;
            this.fileName = handlers.fileName;
            this.config = handlers.config;
            this.ESVersion = handlers.ESVersion;
            this.esGenerator = ECMAScriptLib.createGenerator(handlers.ESVersion);
            this.isWasabyTemplate = handlers.isWasabyTemplate;
         }
         var str = '' + this._process(ast, null, decor);
         if (str) {
            str = '' + str.replace(/\n/g, ' ');
         }
         return JsTemplates.generateTemplate(
            handlers.fileName,
            str,
            handlers.generateTranslations,
            appendHeader,
            this.esGenerator
         );
      },
      getFunction: function getFunction(
         ast,
         data,
         handlers,
         attributes,
         appendHeader
      ) {
         // eslint-disable-next-line no-empty-function
         var func = function () {};
         var str = 'no function';
         try {
            // После аннотации мы знаем имена детей, которые будут находится в _children,
            // эти сведения необходимы в кодогенерации, пробросим их (правда, не самым лучшим образом)
            // до модуля Event
            this.childrenStorage = ast.childrenStorage;

            str = this.getString(ast, data, handlers, attributes, appendHeader);
            func = TemplateLib.createFunction(str);
            func.contentOptionFunctions = this.contentOptionFunctions;
            func.inlineTemplateBodies = this.inlineTemplateBodies;
            func.functionNames = this.functionNames;
            func.internalFunctions = this.internalFunctions;
         } catch (error) {
            errorHandler.info(
               '[Compiler/codegen/function:getFunction()] generating function: \n' +
                  str,
               {
                  fileName: handlers.fileName
               }
            );
            throw error;
         }
         this.setFunctionName(func, undefined, this.fileName);
         this.childrenStorage = [];
         return func;
      },

      /**
       * Генерация подуключаемого модуля
       * @param tag
       * @param data
       * @param decor
       * @returns {Array}
       */
      _processOptionModule: function processOptionModule(tag, data, decor) {
         var tagModule = partialModule.module.call(this, tag, data, decor);
         return tagModule.call(this, decor);
      },

      /**
       * Для поиска модульных функций
       * @param  {Object} tag  Tag
       * @param  {Object} data Data object
       * @param  {Object} decor
       * @return {Object}      Entity: tag or text
       */
      _processModule: function processModule(tag, data, decor) {
         var moduleName = tagUtils.splitWs(tag.name);
         var moduleFunction = this._modules.hasOwnProperty(moduleName)
            ? this._modules[moduleName].module
            : false;
         var tagModule = moduleFunction.call(this, tag, data, decor);
         return tagModule.call(this, decor);
      },

      /**
       * Для загрузки модулей)
       * @param name
       * @returns {Function}
       */
      _processTag: function processTag(name) {
         var modName = tagUtils.splitWs(name);
         if (this._modules[modName] && modName !== 'partial') {
            return this._processModule;
         }
         return this._processEntity;
      },
      _processEntity: function (tag, data, decor, parentNS) {
         if (this._modules[tagUtils.splitWs(tag.name)]) {
            return this._processModule(tag, data, decor);
         }
         if (tagUtils.checkForControl(tag.name, false, false, false)) {
            return this._processOptionModule(tag, data, decor);
         }
         return this._handlingTag(tag, data, decor, parentNS);
      },

      /**
       * Поиск методов шаблонизатора
       * @param  {Object} entity Tag, text, module
       * @return {Function}        Process function
       */
      _whatMethodShouldYouUse: function whatMethodShouldYouUse(entity) {
         var type = utils.capitalize(entity.type),
            functionName = '_process' + type;
         if (
            Array.isArray(this.config.ignored) &&
            this.config.ignored.indexOf(entity.type) > -1
         ) {
            return undefined;
         }
         if (
            entity.type === 'tag' ||
            entity.type === 'style' ||
            entity.type === 'script'
         ) {
            return this._processTag(entity.name);
         }
         if (
            entity.type === 'directive' ||
            entity.type === 'comment' ||
            entity.type === 'text'
         ) {
            return this[functionName];
         }
         return undefined;
      },

      /**
       * Линеаризация AST
       * @param  {Object} entity Tag, text
       * @return {String}
       */
      _stopArrs: function stopArrs(entity) {
         var string = '';
         var i;
         if (Array.isArray(entity)) {
            for (i = 0; i < entity.length; i++) {
               string += entity[i];
            }
            return string;
         }
         return entity;
      },

      /**
       * Поиск модулей в AST
       * @param  {Object} entity Tag, text, module
       * @param  {Object} data   Data object
       * @param prev
       * @param next
       * @param decor
       * @param parentNS
       * @return {String}        Generated string
       */
      _seek: function _seek(entity, data, prev, next, decor, parentNS) {
         var method = this._whatMethodShouldYouUse(entity);
         if (method) {
            entity.prev = prev;
            entity.next = next;
            return this._stopArrs(
               method.call(this, entity, data, decor, parentNS)
            );
         }
         return undefined;
      },

      /**
       * Генерация текст, с переменными
       * @param  {Array} textData Array of data
       * @param  {Object} data     Data
       * @param bindingObject
       * @param wrapUndef
       * @param needEscape
       * @param isAttribute
       * @param attrib
       * @return {String}
       */
      _processData: function processData(
         textData,
         data,
         bindingObject,
         wrapUndef,
         needEscape,
         isAttribute,
         attrib
      ) {
         var string = '';
         var i;
         var expressionResult;
         var result = '';
         if (!textData) {
            throw new Error('Ожидались текстовые данные');
         }
         if (textData.length) {
            for (i = 0; i < textData.length; i++) {
               if (bindingObject) {
                  expressionResult = Process.processExpressions(
                     textData[i],
                     data,
                     this.fileName,
                     bindingObject.isControl,
                     bindingObject.rootConfig,
                     bindingObject.propertyName,
                     isAttribute
                  );
               } else {
                  expressionResult = Process.processExpressions(
                     textData[i],
                     data,
                     this.fileName,
                     undefined,
                     undefined,
                     attrib,
                     isAttribute
                  );
               }

               // todo если считаем значение по умолчанию для биндинга,
               //  и пришел скажем 0 или null, все равно вернется пустая строка.
               if (
                  expressionResult !== undefined &&
                  expressionResult !== null
               ) {
                  if (textData[i].type === 'var') {
                     if (textData[i].localized === true) {
                        string += "' + (" + expressionResult + ") + '";
                     } else {
                        if (wrapUndef) {
                           expressionResult =
                              TClosure.genWrapUndef(expressionResult);
                        }
                        string += "' + (" + expressionResult + ") + '";
                     }
                  } else {
                     string += Process.escapeQuotesInString(expressionResult);
                  }
               } else {
                  string += '';
               }
            }
            result = string;
         } else if (needEscape !== false && textData.type === 'text') {
            result = Process.escapeQuotesInString(textData.value);
         } else if (bindingObject) {
            result = Process.processExpressions(
               textData,
               data,
               this.fileName,
               bindingObject.isControl,
               bindingObject.rootConfig,
               bindingObject.propertyName,
               isAttribute
            );
         } else {
            result = Process.processExpressions(
               textData,
               data,
               this.fileName,
               undefined,
               undefined,
               undefined,
               isAttribute
            );
         }
         if (typeof result === 'string') {
            result = FSC.escapeRawYens(result);
         }
         return result;
      },
      _processAttributesObj: function processAttributesObj(attribs, data, tag) {
         var processed;
         var attrib;
         var obj = {
            attributes: {},
            events: {},
            key: FSC.wrapAroundExec(this.esGenerator.genStringInterpolation(
                this.esGenerator.toTemplateStringExpression('key') + tag.key
            ))
         };
         var eventMeta = {};
         var needEventMeta = true;
         if (attribs) {
            for (attrib in attribs) {
               if (attribs.hasOwnProperty(attrib) && attribs[attrib]) {
                  if (eventExpressions.isEvent(attrib)) {
                     try {
                        var eventObject =
                           eventExpressions.processEventAttribute(
                              attribs[attrib],
                              attrib,
                              data,
                              false,
                              this.fileName,
                              this.childrenStorage,
                              needEventMeta
                           );
                        obj.events[attrib.toLowerCase()] = eventObject.chain;
                        if (needEventMeta) {
                           eventMeta = eventObject.eventMeta;
                           needEventMeta = false;
                        }
                     } catch (error) {
                        throw new Error(
                           'На теге "' +
                              tag.name +
                              '" значение атрибута "' +
                              attrib +
                              '" некорректно "' +
                              attribs[attrib].data[0].name.string +
                              '": ' +
                              error.message
                        );
                     }
                  } else {
                     var isAttribute = true;

                     // todo хак для атрибута data-bind, там не надо эскейпить значение атрибута, потому что могут быть
                     //  кавычки который должны остаться кавычками,
                     //  это выражение позже будет использоваться для привязки данных
                     if (attrib === 'data-bind') {
                        isAttribute = false;
                     }
                     processed = this._processData(
                        attribs[attrib].data,
                        data,
                        undefined,
                        true,
                        false,
                        isAttribute,
                        attrib
                     );
                     var newAttr = attrib.replace('attr:', '');
                     if (utils.removeAllSpaces(processed) !== '') {
                        obj.attributes[newAttr] = processed;
                     } else if (Array.isArray(this.config.booleanAttributes)) {
                        // FIXME: Необходимо проверять все входящие данные перед выполнением сборки шаблона
                        //  Сейчас множественные this, непонятно на что и куда ссылающиеся, мешают. Избавиться от них
                        if (
                           this.config.booleanAttributes.indexOf(
                              attrib.toLowerCase()
                           ) !== -1
                        ) {
                           obj.attributes[newAttr] = 'true';
                        }
                     }
                  }
               }
            }
         }
         if (Object.keys(obj.events).length) {
            obj.events.meta = eventMeta;
            obj.events = FSC.wrapAroundExec(
               'typeof window === "undefined"?{}:' + FSC.getStr(obj.events)
            );
         }
         return obj;
      },

      /**
       * Генерация строки текста
       * @param  {Object} text Text
       * @param  {Object} data Data
       * @return {String}
       */
      _processText: function processText(text, data) {
         var res = this._processData(text.data, data, undefined, true) || '';
         return (
            Generator.genCreateText(
               "'" + res + "'",
                this.esGenerator.genStringInterpolation(
                    this.esGenerator.toTemplateStringExpression('key') + text.key
                )
            ) + ', \n'
         );
      },
      _processDirective: function processDirective(directive) {
         return (
            Generator.genCreateDirective("'" + directive.data + "'") + ', \n'
         );
      },
      _processComment: function processComment(directive, data) {
         var res = this._processData(directive.data, data, undefined, true);
         return Generator.genCreateComment("'" + res + "'") + ', \n';
      },

      /**
       * Генерация строки тэга
       * @param tag
       * @param data
       * @param decor
       * @param parentNS
       * @returns {string}
       */
      _generateTag: function generateTag(tag, data, decor, parentNS) {
         var currentParentNS = parentNS;
         if (tag.attribs && tag.attribs.xmlns) {
            currentParentNS = tag.attribs.xmlns;
         } else if (!tag.attribs) {
            tag.attribs = {
               xmlns: currentParentNS
            };
         } else {
            tag.attribs.xmlns = currentParentNS;
         }
         var attribs =
            typeof decor === 'function' ? decor(tag.attribs) : tag.attribs;
         var processed = this._processAttributesObj(attribs, data, tag);
         Object.keys(processed.attributes).forEach(function (attributeName) {
            processed.attributes[attributeName] = processed.attributes[
               attributeName
            ].replace(/^' \+ (.*?) \+ '$/g, function (str, p) {
               return '¥' + p.replace(/\\/g, '\\\\') + '¥';
            });
         });
         var processedStr = FSC.getStr(processed)
            .replace(/\\("|')/g, '$1')
            .replace(/\\\\/g, '\\')
            .replace(/' \+ /g, '" + ')
            .replace(/ \+ '/g, ' + "');
         var children = this._process(
            tag.children,
            data,
            undefined,
            currentParentNS
         );
         var attrToDecorate = decor
            ? 'attr'
            : 'attr?{context: attr.context, key: ' + this.esGenerator.genStringInterpolation(
             this.esGenerator.toTemplateStringExpression('key') + tag.key
         ) + '}:{}';
         return (
            Generator.genCreateTag(
               "'" + tag.name + "'",
               processedStr,
               children,
               attrToDecorate,
               tag.wsIsContainerNode,
               this.esGenerator
            ) + ', \n'
         );
      },

      /**
       * Генерация тега
       * @param  {Object} tag  Tag
       * @param  {Object} data Array
       * @param decor
       * @param parentNS
       * @return {String}
       */
      _handlingTag: function handlingTag(tag, data, decor, parentNS) {
         return this._generateTag(tag, data, decor, parentNS);
      },

      /**
       * Рекурсивная функция для генерации вёрстки
       * @param  {Array} ast  AST array
       * @param  {Object} data Data
       * @param decor
       * @param parentNS
       * @return {String}
       */
      _process: function process(ast, data, decor, parentNS) {
         var string = '';
         for (var i = 0; i < ast.length; i++) {
            var st = this._seek(
               ast[i],
               data,
               ast[i - 1],
               ast[i + 1],
               decor,
               parentNS
            );
            if (st) {
               string += st;
            }
         }
         return string;
      }
   };
   return processing;
});
