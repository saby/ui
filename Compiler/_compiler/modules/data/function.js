define('Compiler/_compiler/modules/data/function', [
   'Compiler/_compiler/modules/data/utils/functionStringCreator',
   'Compiler/_compiler/modules/utils/parse'
], function (FSC, parseUtils) {
   'use strict';

   /**
    * Для обработки type="function" в конфигурации компонента
    */

   return function functionTag(injected) {
      return FSC.functionTypeHandler(
         this._processData.bind(this),
         injected.children,
         injected.attribs,
         parseUtils.parseAttributesForData
      );
   };
});
