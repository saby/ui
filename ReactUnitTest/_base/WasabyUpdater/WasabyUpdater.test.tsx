/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { delay } from 'Types/function';
import { Control } from 'UICore/Base';

import WasabyUpdaterRoot from '../resources/WasabyUpdater/WasabyUpdaterRoot';

const creator = Control.createControl;
const destroyer = Control.destroyControl;

describe('WasabyUpdater', () => {
    describe('live scenarios', () => {
        const fakeControl = Control.prototype;

        it('basic', () => {
            expect(fakeControl).toBe(fakeControl);
        });
    });

    describe('browser', () => {
        let container: HTMLDivElement;

        beforeEach(() => {
            /*
         _afterMount и _afterUpdate зовутся в отдельной таске, чтобы браузер мог отрисовать кадр.
         Чтобы не делать тесты асинхронными, мы просто мокнем таймеры и сами будем управлять временем.
          */
            jest.useFakeTimers();
            container = document.createElement('div');
            document.body.appendChild(container);
            jest.spyOn(window, 'requestAnimationFrame').mockImplementation(
                setTimeout
            );
        });

        afterEach(() => {
            jest.useRealTimers();
            unmountComponentAtNode(container);
            container.remove();
            container = null;
            jest.restoreAllMocks();
        });

        /**
         * Эта функция существует для того, чтобы не забывать оборачивать тики в act.
         * Это нужно, чтобы реакт реагировал на изменение времени и обновлял компоненты.
         * @param duration Значение, на которое должно продвинуться время.
         */
        function tick(duration: number): void {
            act(() => {
                jest.advanceTimersByTime(duration);
            });
        }

        // В данный момент довольно сложная система вызова обновления, выделим его тик отдельно.
        async function tickWasabyUpdate(): Promise<void> {
            return act(async () => {
                await Promise.resolve();

                // За время requestAnimationFrame отвечает jsdom.
                // Можно было бы с запасом сделать clock.tick(100), но так точнее.
                await new Promise<void>((resolve) => {
                    let resolved: boolean = false;
                    // forceUpdate занимает 2 delay
                    delay(() => {
                        delay(() => {
                            resolved = true;
                            resolve();
                        });
                    });
                    while (!resolved) {
                        jest.advanceTimersByTime(1);
                    }
                });
            });
        }

        it('Очередь на обновление контролов очищается', async () => {
            // region Setup
            let instance;
            act(() => {
                instance = creator(WasabyUpdaterRoot, {}, container);
            });
            tick(0);
            // endregion
            await tickWasabyUpdate();
            const toggleAsync = document.getElementById('toggleAsync');
            const showSync = document.getElementById('showSync');
            act(() => {
                toggleAsync.dispatchEvent(
                    new window.MouseEvent('click', { bubbles: true })
                );
            });
            await tickWasabyUpdate();
            act(() => {
                showSync.dispatchEvent(
                    new window.MouseEvent('click', { bubbles: true })
                );
            });
            tick(2);
            await tickWasabyUpdate();
            expect(container).toMatchSnapshot();

            destroyer(instance, container);
        });
    });
});
