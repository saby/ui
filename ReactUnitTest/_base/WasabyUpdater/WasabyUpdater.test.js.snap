// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WasabyUpdater browser Очередь на обновление контролов очищается 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <div>
      <button
        id="toggleAsync"
      >
        toggleAsync
      </button>
      <button
        id="showSync"
      >
        showSync
      </button>
    </div>
    <div>
      <div
        name="$$wasaby$async$loading"
      />
      <div
        tabindex="0"
      >
        Sync
      </div>
    </div>
  </div>
</div>
`;
